dozer/cogs/filter.py

```diff
@@ -0,0 +1,271 @@
+        if not results:
+            # TODO: Make this a embed
+            await ctx.send("No filters found for this server. Use `{}filter add <>` to add one.".format(
+                ctx.bot.command_prefix))
+            return
+
+        fmt = 'ID {0.id}: `{0.friendly_name}`'
+        if advanced:
+            fmt += ': Pattern: `{0.pattern}`'
+
+        filter_text = '\n'.join(map(fmt.format, results))
+
+        embed = discord.Embed()
+        embed.title = title = "Filters for {}".format(ctx.guild.name)
+        embed.add_field(name="Filters", value=filter_text)
+        await self.check_dm_filter(ctx, embed)
+
+    filter.example_usage = """`{prefix}filter add test` - Adds test as a filter.
+    `{prefix}filter remove 1` - Removes filter 1
+    `{prefix}filter dm true` - Any messages containing a filtered word will be DMed
+    `{prefix}filter whitelist` - See all of the whitelisted roles
+    `{prefix}filter whitelist add Administrators` - Make the Administrators role whitelisted for the filter.
+    `{prefix}filter whitelist remove Moderators` - Make the Moderators role no longer whitelisted."""
+
+    @guild_only()
+    @has_permissions(manage_guild=True)
+    @filter.command()
+    async def add(self, ctx, pattern, friendly_name=None):
+        """Add a pattern to the filter using RegEx. Any word can be added and is tested case-insensitive."""
+        try:
+            re.compile(pattern)
+        except re.error as err:
+            await ctx.send("Invalid ReGex! ```{}```".format(err.msg))
+            return
+        new_filter = WordFilter(guild_id=ctx.guild.id, pattern=pattern, friendly_name=friendly_name or pattern)
+        with db.Session() as session:
+            session.add(new_filter)
+        embed = discord.Embed()
+        embed.title = "Filter {} added".format(new_filter.id)
+        embed.description = "A new filter with the name `{}` was added.".format(friendly_name or pattern)
+        embed.add_field(name="Pattern", value="`{}`".format(pattern))
+        await ctx.send(embed=embed)
+        self.load_filters(ctx.guild.id)
+
+    add.example_usage = "`{prefix}filter whitelist add Swear` - Makes it so that \"Swear\" will be filtered"
+
+    @guild_only()
+    @has_permissions(manage_guild=True)
+    @filter.command()
+    async def remove(self, ctx, filter_id):
+        """Remove a pattern from the filter list."""
+        with db.Session() as session:
+            result = session.query(WordFilter).filter_by(id=filter_id).one_or_none()
+            if result is None:
+                await ctx.send("Filter ID {} not found!".format(filter_id))
+                return
+            if result.guild_id != ctx.guild.id:
+                await ctx.send("That Filter does not belong to this guild.")
+                return
+            result.enabled = False
+            await ctx.send("Filter `{}` with name `{}` deleted.".format(result.id, result.friendly_name))
```

Comment:
This and `self.load_filters(ctx.guild.id)` should be outside the session's `with`-block, so that errors while sending the message or reloading filters don't roll back an otherwise successful transaction. Same applies to `dm_config`.